turbine/src/cluster_nodes.rs

```diff
@@ -395,42 +401,39 @@ fn cmp_nodes_stake(a: &Node, b: &Node) -> Ordering {
 // same TVU socket-addr, we only send shreds to one of them.
 // Additionally limits number of nodes at the same IP address to
 // MAX_NUM_NODES_PER_IP_ADDRESS.
-fn dedup_tvu_addrs(nodes: &mut Vec<Node>) {
+fn dedup_tvu_addrs(nodes: &mut [Node]) {
     const TVU_PROTOCOLS: [Protocol; 2] = [Protocol::UDP, Protocol::QUIC];
     let capacity = nodes.len().saturating_mul(2);
     // Tracks (Protocol, SocketAddr) tuples already observed.
     let mut addrs = HashSet::with_capacity(capacity);
     // Maps IP addresses to number of nodes at that IP address.
     let mut counts = HashMap::with_capacity(capacity);
-    nodes.retain_mut(|node| {
-        let node_stake = node.stake;
+    for node in nodes.iter_mut() {
         let Some(node) = node.contact_info_mut() else {
-            // Need to keep staked identities without gossip ContactInfo for
-            // deterministic shuffle.
-            return node_stake > 0u64;
+            continue;
         };
         // Dedup socket addresses and limit nodes at same IP address.
         for protocol in TVU_PROTOCOLS {
             let Some(addr) = node.tvu(protocol) else {
                 continue;
             };
-            let count: usize = *counts
-                .entry((protocol, addr.ip()))
-                .and_modify(|count| *count += 1)
-                .or_insert(1);
-            if !addrs.insert((protocol, addr)) || count > MAX_NUM_NODES_PER_IP_ADDRESS {
-                // Remove the respective TVU address so that no more shreds are
-                // sent to this socket address.
+            // Check if someone with more stake has claimed the same TVU port before
+            // If so, remove TVU address
+            if !addrs.insert(addr) {
                 node.remove_tvu_addr(protocol);
+            } else {
+                let count: usize = *counts
+                    .entry(addr.ip())
+                    .and_modify(|count| *count += 1)
+                    .or_insert(1);
+                if count > MAX_NUM_NODES_PER_IP_ADDRESS {
+                    // Remove the respective TVU address so that no more shreds are
+                    // sent to the problematic IP
```

Comment:
I don't think this IP is necessarily problematic. It may just mean that a bunch of nodes are directing shreds to some proxy IP

Reply:
this was part of the original logic